### PR TITLE
Kindle: Add wifi selector

### PR DESCRIPTION
**Pull request by notmarek**
https://github.com/koreader/koreader/issues/12040, this is not final, i still need to add https://github.com/notmarek/openlipclua to the base, and i have no idea how to go about it

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/12056)

### CODE REVIEW

File: frontend/device/kindle/device.lua
Comment (NiLuJe) [inline frontend/device/kindle/device.lua:272]: Stray trailing semi-colon (the hazards of mixing C with Lua ;p).

File: frontend/device/kindle/device.lua
Comment (NiLuJe) [inline frontend/device/kindle/device.lua:243]: I don't recall how this all works, but since we do the same on Kobo, I'm going to assume the connectivity check still fires as expected, so you can just drop the commented out code ;).

**Comment by notmarek** [comment]
Found a bug where my lipc handle commits suicide 😅 probably a bug in my underlying library

**Comment by notmarek** [comment]
Whatever magic you did @NiLuJe seems to have fixed it, i assume it has to do with not letting enableWifi abort the connection since that's the only thing i didn't try, thanks!

**Comment by NiLuJe** [comment]
I have a bunch of other changes to test, FWIW, because there's some stupidity left-over from #10669 that ties into that very change ;).

(It's also not really Kindle-specific, I'm pretty sure that scan window leads to funky stuff pretty much everywhere...).

**Comment by notmarek** [comment]
Oh btw for ` password = network.known == "yes" and "HIDDEN",` should i actually pull the password or is that a bad idea?

**Comment by NiLuJe** [comment]
> Oh btw for ` password = network.known == "yes" and "HIDDEN",` should i actually pull the password or is that a bad idea?

I think you're supposed to, yeah. I'm pretty sure we only use that to pre-fill the password input field; and we *will* save that in plaintext, FWIW.

**Comment by NiLuJe** [comment]
WIP dump of the completely untested crap I have in my working tree because I've got to run:

```diff
diff --git a/frontend/device/kindle/device.lua b/frontend/device/kindle/device.lua
index 8542ff7c3..796d4f521 100644
--- a/frontend/device/kindle/device.lua
+++ b/frontend/device/kindle/device.lua
@@ -248,10 +248,7 @@ function Kindle:initNetworkManager(NetworkMgr)
         -- Keep forwarding complete_callback, NetworkMgr:enableWifi will wrap it up in a connectivity check *again*
         -- so it fires *after* isConnect, as the one from reconnectOrShowNetworkMenu itself will be too early:
         -- it's designed for the wpa_supplicant backend, which is blocking, while we're async...
-        -- NOTE: Don't return reconnectOrShowNetworkMenu's status,
-        --       we don't want NetworkMgr:enableWifi to think that the connection failed early,
-        --       as we cannot actually tell (again, we're async, not what the API expects).
-        self:reconnectOrShowNetworkMenu(complete_callback, interactive)
+        return self:reconnectOrShowNetworkMenu(complete_callback, interactive)
     end
 
     function NetworkMgr:turnOffWifi(complete_callback)
diff --git a/frontend/ui/network/manager.lua b/frontend/ui/network/manager.lua
index 826f25851..557f6bc49 100644
--- a/frontend/ui/network/manager.lua
+++ b/frontend/ui/network/manager.lua
@@ -119,6 +119,8 @@ function NetworkMgr:connectivityCheck(iter, callback, widget)
 end
 
 function NetworkMgr:scheduleConnectivityCheck(callback, widget)
+    print("NetworkMgr:scheduleConnectivityCheck")
+    print(debug.traceback())
     self.pending_connectivity_check = true
     UIManager:scheduleIn(0.25, self.connectivityCheck, self, 1, callback, widget)
 end
@@ -175,8 +177,8 @@ end
 function NetworkMgr:getNetworkInterfaceName() end
 function NetworkMgr:getNetworkList() end
 function NetworkMgr:getCurrentNetwork() end
-function NetworkMgr:authenticateNetwork() end
-function NetworkMgr:disconnectNetwork() end
+function NetworkMgr:authenticateNetwork(network) end
+function NetworkMgr:disconnectNetwork(network) end
 -- NOTE: This is currently only called on hasWifiManager platforms!
 function NetworkMgr:obtainIP() end
 function NetworkMgr:releaseIP() end
@@ -317,8 +319,23 @@ function NetworkMgr:canResolveHostnames()
 end
 
 -- Wrappers around turnOnWifi & turnOffWifi with proper Event signaling
-function NetworkMgr:enableWifi(wifi_cb, connectivity_cb, connectivity_widget, interactive)
-    local status = self:requestToTurnOnWifi(wifi_cb, interactive)
+function NetworkMgr:enableWifi(wifi_cb, interactive)
+    local connectivity_cb
+    -- NOTE: If we're interactive, let the backend run the connectivity check once it's actually attempted a connection,
+    --       this prevents reconnectOrShowNetworkMenu from having a kill switch ticking behind its back while the scan widget is up,
+    --       which is what would happen if *we* were to launch the connectivity check...
+    if interactive then
+        connectivity_cb = function()
+            -- This will handle sending the proper Event, manage wifi_was_on, as well as tearing down Wi-Fi in case of failures.
+            self:scheduleConnectivityCheck(wifi_cb)
+        end
+    else
+        connectivity_cb = wifi_cb
+    end
+    -- FIXME: Fix android & sdl turnOnWifi so that they actually honor complete_callback...
+    --        ... probably via a scheduleConnectivityCheck ;p.
+
+    local status = self:requestToTurnOnWifi(connectivity_cb, interactive)
     -- If turnOnWifi failed, abort early
     if status == false then
         logger.warn("NetworkMgr:enableWifi: Connection failed!")
@@ -330,42 +347,30 @@ function NetworkMgr:enableWifi(wifi_cb, connectivity_cb, connectivity_widget, in
         -- If it wasn't, it might have been from a beforeWifiAction, and, much like in turnOnWifiAndWaitForConnection,
         -- we don't want to risk rescheduling the same thing over and over again.
         if interactive then
-            -- Unlike the next branch, turnOnWifi was *not* called, so we don't need the extra checks.
-            self:scheduleConnectivityCheck(connectivity_cb, connectivity_widget)
-        else
-            -- No connectivity check to handle that for us, so close the widget *now*
-            if connectivity_widget then
-                UIManager:close(connectivity_widget)
-            end
+            -- Unlike the next branch, turnOnWifi was *not* called, so we don't need the extra guards around concurrent checks.
+            -- (This is obviously the exact same thing as what our connectivity_cb wrapper would do, but make it explicit for clarity).
+            self:scheduleConnectivityCheck(wifi_cb) -- i.e., connectivity_cb()
         end
         return
     else
-        -- Some turnOnWifi implementations may fire a connectivity check,
-        -- but we *need* our own, because of the callback & widget passing,
-        -- as we might have been called by the "prompt" beforeWifiAction...
-        -- NOTE: We *could* arguably have multiple connectivity checks running concurrently,
-        --       but only having a single one running makes things somewhat easier to follow...
-        -- NOTE: Also, most of the platforms that use a connectivity check in turnOnWifi do it to handle wifi_cb,
-        --       so we'll want to preserve it by wrapping both possible callbacks in a single function...
-        local wrapped_cb
-        if self.pending_connectivity_check then
-            self:unscheduleConnectivityCheck()
-
-            wrapped_cb = function()
-                if wifi_cb then
-                    wifi_cb()
-                end
-                if connectivity_cb then
-                    connectivity_cb()
-                end
+        -- If we're interactive, we've already handled the connectivity check & cb via our backend
+        if not interactive then
+            -- Some turnOnWifi implementations may fire a connectivity check,
+            -- but we *need* our own, because of the callback & widget passing,
+            -- as we might have been called by the "prompt" beforeWifiAction...
+            -- NOTE: We *could* arguably have multiple connectivity checks running concurrently,
+            --       but only having a single one running makes things somewhat easier to follow...
+            -- NOTE: Most of the platforms that use a connectivity check in turnOnWifi do it to handle
+            --       turnOnWifi's complete_callback, i.e., our own wifi_cb (usually because of an async backend).
+            --       Which means that if there *isn't* a connectivity check scheduled,
+            --       we can assume that wifi_cb already ran, via requestToTurnOnWifi,
+            --       most likely because the backend is our own, and reconnectOrShowNetworkMenu handled it ;).
+            if self.pending_connectivity_check then
+                self:unscheduleConnectivityCheck()
             end
-        else
-            -- If the turnOnWifi implementation didn't rely on a connectivity check, assume wifi_cb was already run.
-            wrapped_cb = connectivity_cb
-        end
 
-        -- This will handle sending the proper Event, manage wifi_was_on, as well as tearing down Wi-Fi in case of failures.
-        self:scheduleConnectivityCheck(wrapped_cb, connectivity_widget)
+            self:scheduleConnectivityCheck(wifi_cb)
+        end
     end
 
     return true
@@ -396,7 +401,7 @@ function NetworkMgr:toggleWifiOn(complete_callback, long_press, interactive)
 
     self.wifi_toggle_long_press = long_press
 
-    self:enableWifi(complete_callback, nil, nil, interactive)
+    self:enableWifi(complete_callback, interactive)
 
     UIManager:close(toggle_im)
 end
@@ -831,6 +836,8 @@ function NetworkMgr:getWifiToggleMenuTable()
         self:queryNetworkState()
         local fully_connected = self.is_wifi_on and self.is_connected
         local complete_callback = function()
+            print("Running NetworkMgr complete_callback on", touchmenu_instance)
+            print(debug.traceback())
             -- Notify TouchMenu to update item check state
             touchmenu_instance:updateItems()
         end -- complete_callback()
@@ -1002,7 +1009,7 @@ function NetworkMgr:getDismissScanMenuTable()
     return {
         text = _("Dismiss Wi-Fi scan popup after connection"),
         checked_func = function() return G_reader_settings:nilOrTrue("auto_dismiss_wifi_scan") end,
-        enabled_func = function() return Device:hasWifiManager() and not Device:isEmulator() end,
+        enabled_func = function() return Device:hasWifiManager() or Device:isEmulator() end,
         callback = function() G_reader_settings:flipNilOrTrue("auto_dismiss_wifi_scan") end,
     }
 end
@@ -1041,7 +1048,7 @@ function NetworkMgr:reconnectOrShowNetworkMenu(complete_callback, interactive)
     UIManager:close(info)
     if network_list == nil then
         UIManager:show(InfoMessage:new{text = err})
-        return
+        return false
     end
     -- NOTE: Fairly hackish workaround for #4387,
     --       rescan if the first scan appeared to yield an empty list.
@@ -1051,14 +1058,16 @@ function NetworkMgr:reconnectOrShowNetworkMenu(complete_callback, interactive)
         network_list, err = self:getNetworkList()
         if network_list == nil then
             UIManager:show(InfoMessage:new{text = err})
-            return
+            return false
         end
     end
 
     table.sort(network_list,
         function(l, r) return l.signal_quality > r.signal_quality end)
 
-    local success = false
+    -- true: we're connected; false: things went kablooey; nil: we don't know yet (e.g., interactive)
+    -- NOTE: false *will* lead enableWifi to kill Wi-Fi via _abortWifiConnection!
+    local success
     if self.wifi_toggle_long_press then
         self.wifi_toggle_long_press = nil
     else
@@ -1120,6 +1129,9 @@ function NetworkMgr:reconnectOrShowNetworkMenu(complete_callback, interactive)
                 network_list = network_list,
                 connect_callback = complete_callback,
             })
+        else
+            -- Let enableWifi tear it all down when we're non-interactive
+            success = false
         end
     end
 
diff --git a/frontend/ui/network/networklistener.lua b/frontend/ui/network/networklistener.lua
index 2e783eaef..3a0d0cf1f 100644
--- a/frontend/ui/network/networklistener.lua
+++ b/frontend/ui/network/networklistener.lua
@@ -30,7 +30,7 @@ local function enableWifi()
     -- NB Normal widgets should use NetworkMgr:promptWifiOn()
     -- (or, better yet, the NetworkMgr:beforeWifiAction wrappers: NetworkMgr:runWhenOnline() & co.)
     -- This is specifically the toggle Wi-Fi action, so consent is implied.
-    NetworkMgr:enableWifi(nil, nil, nil, true) -- flag it as interactive
+    NetworkMgr:enableWifi(nil, true) -- flag it as interactive
 
     UIManager:close(toggle_im)
 end
```

(Specifically, what I mentioned earlier about #10669 are the completely mystifying `connectivity_` args to `enableWifi` that were never ever used (not even in *any* commits during that PR's testing). I have no idea what I was thinking ;D).

**Comment by NiLuJe** [comment]
(I have yet to test the latest NetworkMgr cleanups, but I need a nap ;p)

**Comment by NiLuJe** [comment]
(nap time, for real ;p)

**Comment by notmarek** [comment]
Yay lipc errors are back ;)

**Comment by NiLuJe** [comment]
Okay, dealt with some quirks on old FW (no netid field ;'().

Also, sorry, I kinda ruined your scheduling for the scan and just made it block like the wpa backend ;).

There wasn't anything wrong with that approach, mind you, except that it essentially prevented early errors from being caught, well, early, and made the various "enabling/scanning/connected" infomessages effectively fire in the same frame.
(They're still fairly close together on devices with fast scanning (i.e., no 5GHz), though).

Surprisingly enough, last night's sleep-deprived hacking around didn't horribly break all the things. I'll double-check on Kobo, but so far it looks okay ;).

**Comment by NiLuJe** [comment]
I should also have dealt with most of the quirks involved in dealing with the scan list widgets, which was full of weird edge-cases because nobody really uses it ;p.

Which reminds me: there's no `disconnectNetwork` implementation, so the "disconnect" button in the scan list doesn't do anything.

Not sure if we can even *do* something about that, or if the fact that connecting to something else works anyway is good enough?

**Comment by NiLuJe** [comment]
Also, FWIW, my `current_profile` is always `""` by the time the scan is done on that device. Not sure if you're seeing anything else on newer devices with the netid stuff (and/or slower scans because 5Ghz).

**Comment by notmarek** [comment]
Yeah my profile is usually the saved network when I scan on the latest fw
on 11th gen kindle

**Comment by NiLuJe** [comment]
Cool, I'd be happy to know if the `connected` flag gets set properly in `getNetworkList` on your end, then.

Which means that `NetworkMgr:reconnectOrShowNetworkMenu` should catch it in its first pass and *not* call `authenticateNetwork` manually.

And that hopefully this leads to stuff that works ;D.

**Comment by NiLuJe** [comment]
Oh, speaking of passes, was the two-pass `kindleGetScanList` necessary for some reason on your end?

(There's a similar workaround in the front-end, but it's because of a weird bug in a very specific old Kobo FW on a specific device; so just curious if you actually encountered an issue, or if it was just to replicate that ;)).

**Comment by notmarek** [comment]
Everything broke after you pushed the changes this was only way to make it
stable😅

**Comment by NiLuJe** [comment]
Ahaha :D.

Well, guess you'll have to test whether the latest round of "fixes" work on your end, because that PW2 really really doesn't want to linger scanning, so I basically never ever wait for it...

```
   240623:005221 wifid[12477]: I wmgr:scan:hidden-nets=0 :starting...::~:
   240623:005221 wifid[12477]: I spectator:scan-start:t=50862.520761:
   240623:005221 wifid[12477]: I spectator:scan-complete:t=50862.881703:
```

**Comment by NiLuJe** [comment]
On the upside, stuff looks sounds on Kobo, so, yay, didn't break *all* the things ;p

**Comment by notmarek** [comment]
> On the upside, stuff looks sound on Kobo, so, yay, didn't break _all_ the things ;p

seems to works unless i spam click the connection button, which then gives me a scan results == nil error

**Comment by NiLuJe** [comment]
> seems to works unless i spam click the connection button, which then gives me a scan results == nil error

Huh. I can't repro on a that device (probably because of the instant scans, again), but I tweaked the "in between" popup to be more explicit and less foot-gun-y if there's actually a connection attempt in progress.

(Which should always be the case if you see this popup on Kindle, unlike on other platforms (e.g., on most Kobo, if the driver is loaded but no if is up, which *can* happen depending on the previous state of the system before we launched).

File: frontend/device/kindle/device.lua
Comment (notmarek) [inline frontend/device/kindle/device.lua:1]: Should i do this? kindleGetCurrentProfile basically does this already, just needs to translate the date the same way we do for the scanList :)

**Comment by NiLuJe** [comment]
I *think* I caught your nil shenanigans with [e34d3aa](https://github.com/koreader/koreader/pull/12056/commits/e34d3aafe65f47ec9daaadc3b773a7f1d08af12b) (or at least something similar).

Dealing with NULLs in C objects is weird (whether they be cdata or udata) ;).

File: frontend/device/kindle/device.lua
Comment (NiLuJe) [reply frontend/device/kindle/device.lua:1]: Sure, go ahead ;).

That's one less potentially weird issue down the line, if this can actually ever happen ;).

**Comment by notmarek** [comment]
```
 [*] Current time: 06/23/24-19:52:00
 [*] Version: v2024.04-196-gfa4654cdf_2024-06-19

ffi.load: libs/libutf8proc.so.3
ffi.load: blitbuffer
ffi.load: sqlite3
ffi.load: fbink_input
ffi.load (assisted searchpath): ./libs/libfbink_input.so.1
[FBInk] /dev/input/event0: `bd71828-pwrkey` = KEY | POWER_BUTTON
[FBInk] /dev/input/event1: `pt_mt` = TOUCHSCREEN
[ko-input] Forked off fake event generator (pid: 24053)
06/23/24-19:52:01 INFO  initializing for device KindleBasic4
06/23/24-19:52:01 INFO  framebuffer resolution: {
  h = 1448,
  w = 1072
} --[[table: 0xb6b3d3b8]]
ffi.load: libs/libmupdf.so
ffi.load: libs/libwrap-mupdf.so
ffi.load: libs/libfreetype.so.6
ffi.load: libs/libharfbuzz.so.0
ffi.load: libs/libzstd.so.1
06/23/24-19:52:02 INFO  opening file /mnt/us/Stoka/Sanderson, Brandon - [The Stormlight Archive 3.5] Dawnshard [61689] (r1.0) [EN].epub
CRE: font NotoSerifKR regular: updated weight from 500 to 400
CRE: font NotoSerifKR bold: updated weight from 900 to 700
CRE: font PMN Caecilia LT regular: updated weight from 500 to 400
CRE: font NotoSansKR bold: updated weight from 900 to 700
06/23/24-19:52:02 INFO  Inhibiting user input
06/23/24-19:52:02 INFO  Loading plugins from directory: plugins
06/23/24-19:52:04 INFO  Restoring user input handling
06/23/24-19:52:24 INFO  Wi-Fi successfully restored (after 2 seconds)!
06/23/24-19:52:38 WARN  NetworkMgr:enableWifi: Connection failed!
06/23/24-19:52:47 WARN  NetworkMgr:enableWifi: Connection failed!
```
This is all i get when i get nil response from scanlist, then the whole thing dies (as in i need to force a reboot)

**Comment by NiLuJe** [comment]
Ah, no, I got a "nicer" crash on my end ;p.

```
./luajit: frontend/device/kindle/device.lua:376: bad argument #1 to 'ipairs' (table expected, got string)
stack traceback:
        [C]: in function 'ipairs'
        frontend/device/kindle/device.lua:376: in function 'getNetworkList'
        frontend/ui/network/manager.lua:1056: in function 'requestToTurnOnWifi'
        frontend/ui/network/manager.lua:338: in function 'enableWifi'
        frontend/ui/network/manager.lua:405: in function 'toggleWifiOn'
        frontend/ui/network/manager.lua:859: in function 'callback'
        frontend/ui/widget/touchmenu.lua:922: in function 'onMenuSelect'
        frontend/ui/widget/touchmenu.lua:216: in function 'handleEvent'
        frontend/ui/widget/container/inputcontainer.lua:266: in function 'handleEvent'
        frontend/ui/widget/container/widgetcontainer.lua:83: in function 'propagateEvent'
        frontend/ui/widget/container/widgetcontainer.lua:101: in function 'handleEvent'
        ...
        frontend/ui/widget/container/widgetcontainer.lua:101: in function 'handleEvent'
        frontend/ui/widget/container/widgetcontainer.lua:83: in function 'propagateEvent'
        frontend/ui/widget/container/widgetcontainer.lua:101: in function 'handleEvent'
        frontend/ui/uimanager.lua:915: in function 'sendEvent'
        frontend/ui/uimanager.lua:53: in function '__default__'
        frontend/ui/uimanager.lua:1446: in function 'handleInputEvent'
        frontend/ui/uimanager.lua:1546: in function 'handleInput'
        frontend/ui/uimanager.lua:1590: in function 'run'
        ./reader.lua:280: in main chunk
```

**Comment by notmarek** [comment]
Seems to be stable now ;)

**Comment by notmarek** [comment]
> ./luajit: frontend/device/kindle/device.lua:376: bad argument #1 to 'ipairs' (table expected, got string)
> stack traceback:
>         [C]: in function 'ipairs'
>         frontend/device/kindle/device.lua:376: in function 'getNetworkList'
>         frontend/ui/network/manager.lua:1056: in function 'requestToTurnOnWifi'
>         frontend/ui/network/manager.lua:338: in function 'enableWifi'
>         frontend/ui/network/manager.lua:405: in function 'toggleWifiOn'
>         frontend/ui/network/manager.lua:859: in function 'callback'
>         frontend/ui/widget/touchmenu.lua:922: in function 'onMenuSelect'
>         frontend/ui/widget/touchmenu.lua:216: in function 'handleEvent'
>         frontend/ui/widget/container/inputcontainer.lua:266: in function 'handleEvent'
>         frontend/ui/widget/container/widgetcontainer.lua:83: in function 'propagateEvent'
>         frontend/ui/widget/container/widgetcontainer.lua:101: in function 'handleEvent'
>         ...
>         frontend/ui/widget/container/widgetcontainer.lua:101: in function 'handleEvent'
>         frontend/ui/widget/container/widgetcontainer.lua:83: in function 'propagateEvent'
>         frontend/ui/widget/container/widgetcontainer.lua:101: in function 'handleEvent'
>         frontend/ui/uimanager.lua:915: in function 'sendEvent'
>         frontend/ui/uimanager.lua:53: in function '__default__'
>         frontend/ui/uimanager.lua:1446: in function 'handleInputEvent'
>         frontend/ui/uimanager.lua:1546: in function 'handleInput'
>         frontend/ui/uimanager.lua:1590: in function 'run'
>         ./reader.lua:280: in main chunk

This hasnt been happening to me since you update openlipclua to return nil

**Comment by NiLuJe** [comment]
FWIW, I can repro the above if I ninja murder Wi-Fi *during* a connection attempt (e.g., via the aforementioned "in-between" popup you get from the "Wi-Fi connection" menu entry).

Now that it's handle,d you get a "Cannot access scan results" error instead of a crash, but it persists for a while until wifid gets its bearings again.

(The syslog concurs that shit's going wonky, e.g., I see scan timeouts and shit).

**Comment by NiLuJe** [comment]
> Now that it's handled, you get a "Cannot access scan results" error popup instead of a crash, but it persists for a while until wifid gets its bearings again.

Hadn't seen your commit yet, returning an empty list is much better ;).

So, now, it trips a rescan in NetworkMgr, and shit just works with a slight delay ;).

**Comment by notmarek** [comment]
> > Now that it's handled, you get a "Cannot access scan results" error popup instead of a crash, but it persists for a while until wifid gets its bearings again.
> 
> Hadn't seen your commit yet, returning an empty list is much better ;).
> 
> So, now, it trips a rescan in NetworkMgr, and shit just works with a slight delay ;).

Yeah i thought it would work better after reading your comment about the error actually cancelling everything

**Comment by NiLuJe** [comment]
Makes you wanna think about what `openlipcluaha_to_table` should actually return on error, though, as a string is kind of unexpected ;p.

Throw an error? A `nil`? An empty table?

**Comment by notmarek** [comment]
> Makes you wanna think about what `openlipcluaha_to_table` should actually return on error, though, as a string is kind of unexpected ;p.
> 
> Throw an error? A `nil`? An empty table?

nil would probably be the sanest option

**Comment by NiLuJe** [comment]
> nil would probably be the sanest option, but then again an empty table seems like a good choice too

On it!

**Comment by NiLuJe** [comment]
(Commit msg for the squash)

* Kindle: Implement a NetworkMgr backend loosely based on WpaClient in order to allow feature-parity with hasWifiManager platforms. This involves dealing with the native wifid over lipc (the native IPC system, based on DBus), through custom Lua bindings (https://github.com/notmarek/openlipclua), since the stock ones lack support for the needed hasharray data type.
* NetworkMgr: Clear up leftover hallucinations from #10669, making `enableWifi` much simpler (and much more similar to `turnOnWifiAndWaitForConnection`).
* NetworkMgr: Made it clearer that `turnOnWifi` implementations *must* deal with `complete_callback`, as part of the aforementioned changes mean that it's *always* wrapped in a connectivity check, and we need that for proper event signaling.
* Android, Emu: Run `complete_callback` properly in `turnOnWifi`.
* Kindle: Support `powerd:isCharged()` on the PW2 (yes, this is random, it just happened to be my test device :D).
* NetworkMgr:disableWifi: Properly tear down any potential ongoing connection attempt (e.g., connectivity check).
* NetworkMgr:promptWifi: Make the "wifi enabled but not connected" popup clearer if there's an ongoing connection attempt, and gray out the "Connect" button in this case (as it would only lead to another "connection already in progress" popup anyway).
* NetworkMgr:reconnectOrShowNetworkMenu: Make *total* scanning failures fatal (they will lead to an immediate wifi teardown).
* NetworkMgr:reconnectOrShowNetworkMenu: Clear up the long-press behavior (which *always* shows the network list popup) so that it doesn't weirdly break all the things (technical term!).
* NetworkMgr:reconnectOrShowNetworkMenu: When we manage to connect to a preferred network on our own *before* showing the network list, make sure it is flagged as "connected" in said list.
* NetworkMgr:reconnectOrShowNetworkMenu: Make connection failures fatal in non-interactive workflows (they'll lead to a wifi teardown).
* NetworkSetting (the aforementioned network list widget): Clear NetworkMgr's "connection pending" flag on dismiss when nothing else will (i.e., when there's no connectivity check ticking).

**Review by NiLuJe — APPROVED** [review]
LGTM! (Needs a base bump, though)

> Reviewed 7 of 7 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12056#-:-O06DXS32WwxcaTYVr_w:b413lc7)* status: all files reviewed, 1 unresolved discussion (waiting on @notmarek)

**Review by notmarek — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12056#-:-O06I7Ui5j4OInCMeokt:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @notmarek)

**Comment by benoit-pierre** [comment]
It would be better to have a preliminary base bump (up to just before https://github.com/benoit-pierre/koreader-base/commit/a9e32ffc7c6ba1644a788f59fc228ba4d056571f) in a separate PR, and after it's merged you rebase with a bump to https://github.com/benoit-pierre/koreader-base/commit/a9e32ffc7c6ba1644a788f59fc228ba4d056571f only. This way it's only related changes.

I imagine this will be squashed on merge, otherwise, I'd also move the bump as the first commit of the series (since the rest of the stuff depend on it).

**Comment by NiLuJe** [comment]
There's also a question of whether we wait for after the release because there's a few new translatable string ;).

Thoughts, @Frenzie?

**Comment by Frenzie** [comment]
Fine to merge, however in that case I haven't reviewed it yet.

File: frontend/device/kindle/device.lua
Comment (Frenzie) [inline frontend/device/kindle/device.lua:123]: If we want to show something so technical it probably needs a translator's note.

File: frontend/device/kindle/device.lua
Comment (Frenzie) [inline frontend/device/kindle/device.lua:135]: Same comment for anything with lipc, in this case also for NetworkMgr which shouldn't be altered.

File: frontend/device/kindle/device.lua
Comment (Frenzie) [inline frontend/device/kindle/device.lua:183]: I'd say one of these.
```suggestion
        return nil, _("Timed out while scanning for Wi-Fi networks")
```

```suggestion
        return nil, _("Scanning for Wi-Fi networks timed out")
```

File: frontend/device/kindle/device.lua
Comment (Frenzie) [reply frontend/device/kindle/device.lua:124]: I think wifi backend sounds better.

**Review by Frenzie — APPROVED** [review]

**Review by notmarek — COMMENTED** [review]
> Reviewed 5 of 7 files at r2, 2 of 2 files at r3, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12056)* status: all files reviewed, 4 unresolved discussions (waiting on @NiLuJe)

**Comment by NiLuJe** [comment]
(Commit history archived in the [kindle-wifi-ui-openlipc](https://github.com/NiLuJe/koreader/tree/kindle-wifi-ui-openlipc) branch ;).

**Comment by mergen3107** [comment]
@notmarek 
Is this UI supposed to show up only when there were no network connections?
I have connection to eduroam network, and I don't know how to "forget" it in Kindle Menu.

Also, this one is a simple username/password UI?
eduroam needs all kinds of parameters for Enterprise settings, so this PR is probably won't work for me :D

**Comment by NiLuJe** [comment]
It's the standard UI, and behaves roughly like the stock ones, i.e., unless it can connect to a previously setup preferred netwokr, you get shown a network list.

And, indeed, eduroam would require manual labor to setup ;).

(My only experience with eduroam was quite a few years ago on macOS, so things may be slightly less ridiculous nowadays. But, FWIW, I'd never done that on an eReader, it was usually faster to setup a Wi-Fi hotspot on my phone).

**Comment by mergen3107** [comment]
> unless it can connect to a previously setup preferred netwokr, you get shown a network list.

Can I see network list even with existing network?

I noticed I have a very weird thing. Kindle connects to a different network that I probably accidentally hit one day. 

While in KOReader, I turn on Wi-Fi, then it connects to this stupid network and I need to quit and setup eduroam all over again :D 

This PR could have helped me save a few steps. 

I vaguely remember something about /var/local/ folder and wpa-supplicant files I needed to wipe to forget all networks. Are there any gentler ways to forget networks on Kindle? :D

**Comment by NiLuJe** [comment]
A long-press on the Wi-Fi menu entry will force the scan list to display (it *won't* prevent the connection attempt to go through, though, so you'll see the preferred network as "connected". Even though, on Kindle, it's sort of not yet actually connected yet ^^).

I imagine you can forget about networks in the stock WiFi setting, though.

**Comment by mergen3107** [comment]
> A long-press on the Wi-Fi menu entry will force the scan list to display
Is this in KOReader?

**Comment by mergen3107** [comment]
Lol, I figured out how to forget networks in Kindle. After it is connected, hit it again, and it will ask to forget the network xD

**Comment by mergen3107** [comment]
I found a cosmetic bug.
I was at home, connected to home network OK.
Then went to school with eduroam, and it said that it is connecting to home network, but actually connected to eduroam OK.

**Comment by notmarek** [comment]
probably a kindle bug since the info is grabbed straight from lipc

**Comment by mergen3107** [comment]
I still can't find a way to bring up a list of networks (Kindle Scribe).
Long pressing on the `Wi-Fi Connection` brings up a popup window that says
```
Wi-Fi is enabled, but you're currently not connected to a network.
How would you like to proceed?
Dismiss
Turn Wi-Fi off
Connect
```
However, that statement is wrong. Kindle is connected and I was able to update over the OTA and connected via sftp to PC.

**Comment by notmarek** [comment]
I don't think there's a forget button in koreader, otherwise I would've
implemented it too

**Comment by mergen3107** [comment]
> probably a kindle bug since the info is grabbed straight from lipc

@notmarek 
Should I play around with pause times in between commands?

**Comment by mergen3107** [comment]
> However, that statement is wrong.

Oops, it seems network actually did not connect the second time :D

**Comment by mergen3107** [comment]
OK, so some mysteries. I don't remember if that starting happening with this PR though.
Sorry for the spam (I can also start a new ticket if you guys insist).
1) KOReader connects to eduroam first time in a day no problem.
2) However, if I try connect and disconnect in a short time in between it couldn't connect anymore and I need to go Kindle's menu to reconfigure it again (account and password after a 2-min timeout, mainly).
3) In some of those time when eduroam connection is failed KOReader connects to a different password-less network, which I never connected to before with Kindle at all.
Does KOReader fallback to that network or something?

**Comment by NiLuJe** [comment]
We don't do *anything*, it's all wifid.

**Pull request merged**
